Initial specification of JDQL (#520)
jakartaee/data: spec/src/main/asciidoc/query-language.asciidoc

- operators and punctuation characters,
- literal strings, and
- integer and decimal number literals.

==== Identifiers and keywords

An _identifier_ is any legal Java identifier which is not a keyword. Identifiers are case-sensitive: `hello`, `Hello`, and `HELLO` are distinct identifiers.

In the JDQL grammar, identifiers are labelled with the `IDENTIFIER` token type.

The following identifiers are _keywords_: `select`, `update`, `set`, `delete`, `from`, `where`, `order`, `by`, `asc`, `desc`, `not`, `and`, `or`, `between`, `like`, `in`, `null`, `local`, `true`, `false`. In addition, every reserved identifier listed in section 4.4.1 of the Jakarta Persistence specification version 3.2 is also considered a reserved identifier. Keywords and other reserved identifiers are case-insensitive: `null`, `Null`, and `NULL` are three ways to write the same keyword.

NOTE: Use of a reserved identifier as a regular identifier in JDQL might be accepted by a given Jakarta Data provider, but such usage is not guaranteed to be portable between providers.

==== Parameters

A _named parameter_ is a legal Java identifier prefixed with the `:` character, for example, `:name`.

> An _ordinal parameter_ is a decimal integer prefixed with the `?` character, for example, `?1`.

Reviewer: It would be hard inclusion of `@`.
This is more popular this use at NoSQL databases.

When executed, an enum expression evaluates to the named member of the Java `enum` type.

==== Path expressions

A _path expression_ is a period-separated list of Java identifiers, with syntax specified by `state_field_path_expression`. Each identifier is interpreted as the name of a field of an entity or embeddable class. Each prefix of the list is assigned a Java type:

- the first element of the list is assigned the type of the named field of the entity being queried, and
- each subsequent element is assigned the type of the named field of the type assigned to the previous element.

The type of the whole path expression is the type of the last element of the list. For example, `pages` is assigned the type `int`, `address` is assigned the type `org.example.Address`, and `address.street` is assigned the type `java.lang.String`.

When executed, a path expression is evaluated in the context of a given record of the queried entity type, and evaluates to the value of the entity field for the given record.

> ==== Function calls

Reviewer: About those functions, there is no guarantee of this function support on the NoSQL database side.

Author: Yeah, so what I want to know is which functions _can_ be implemented, "typically"? Or at least, if not "typically", which can be implemented for a sufficient subset of platforms that it's worth mentioning them here/

Author: I mean, if the answer is simply that none of these things can typically be implemented on non-relational databases, then we should simply remove this section.

Reviewer: Typically at Document and Graph:

| Function | Supported by | Not Supported by |
|----------|--------------|------------------|
| `abs()`  | Document, Graph | Wide-column, Key-value |
| `length()`| Document, Graph | Wide-column, Key-value |
| `lower()` | Document, Graph | Wide-column, Key-value |
| `upper()` | Document, Graph | Wide-column, Key-value |
| `left()`  | Document, Graph | Wide-column, Key-value |
| `right()` | Document, Graph | Wide-column, Key-value |

But even with document, I cannot guarantee, e.g.:

- ravendb
- harperdb

Reviewer: All functions, and an UnsupportedOperationException is thrown only when the database provider does not support it.

Author: All the functions on that list?

Reviewer: Yes!

- abs
- length
- lower
- upper
- left
- right

Author: Can `local date` and `local datetime` be supported on all platforms? (I'm guessing "no", but looking for confirmation.)

Author: Please see commit https://github.com/jakartaee/data/pull/520/commits/858686efec3b0f261d464851156e3d58fad4037c.

Reviewer 2: can we mandate compilation time checking of supported functions? runtime errors are not nice

Author: Well the issue is that at the spec level, Jakarta Data allows purely-runtime implementations.

Now, as far as I can tell, we're all implementing the spec using annotation processing, and so, indeed, we can all check this at compilation time (I certainly do).

But if we were to _require_ that, then we would be prohibiting other implementation strategies.

Author: As discussed on a recent concall, I already added `local time`. And it seems we have a better-than-tentative consensus on the proposed list of functions.

When executed, an enum expression evaluates to the named member of the Java `enum` type.

==== Path expressions

A _path expression_ is a period-separated list of Java identifiers, with syntax specified by `state_field_path_expression`. Each identifier is interpreted as the name of a field of an entity or embeddable class. Each prefix of the list is assigned a Java type:

- the first element of the list is assigned the type of the named field of the entity being queried, and
- each subsequent element is assigned the type of the named field of the type assigned to the previous element.

The type of the whole path expression is the type of the last element of the list. For example, `pages` is assigned the type `int`, `address` is assigned the type `org.example.Address`, and `address.street` is assigned the type `java.lang.String`.

> When executed, a path expression is evaluated in the context of a given record of the queried entity type, and evaluates to the value of the entity field for the given record.

Reviewer 2: what happens when something in the path is `null`? Do we support `null` safe dereferencing?

Author: This is something I need to clarify.

Author: @graemerocher see 34ca9693c2bb8fce5c2b3e9c84dd69644de008f0.

The `from` clause is optional in `select` statements. When it is missing, the queried entity is determined by the return type of the repository method, or, if the return type is not an entity type, by the primary entity type of the repository.

NOTE: The syntax of the `update` statement is irregular, with the `from` keyword implied. That is, the syntax _should_ be `update from Entity`, but for historical reasons it is simply `update Entity`.

==== Where clause

The `where` clause, with syntax given by `where_clause`, specifies a conditional expression used to restrict the records returned, deleted, or updated by the query. Only records for which the conditional expression is satisfied are returned, deleted, or updated.

The `where` clause is always optional. When it is missing, there is no restriction, and, and all records are of the queried entity type are returned, deleted, or updated.

==== Select clause

> The `select` clause, with syntax given by `select_clause`, specifies a list of expressions which are returned by the query. Each expression in the list is evaluated for each record which satisfies the restriction imposed by the `where` clause. Alternatively, the `select` clause may contain a single `count(*)` aggregate expression, which evaluates to the number of records which satisfy the restriction.

Reviewer 2: what happens in the db doesn't support `count(*)`?

Reviewer: I guess we put the Unsupportedoperationexception, am I right, @njr-11 ?

Author: @graemerocher @otaviojava please review d127e7731db03a209101278ace53c2cf13933d41

Reviewer 3: > I guess we put the Unsupportedoperationexception, am I right, @njr-11 ?

Right, based on how we treat other unsupported function in the spec, it should be `UnsupportedOperationException`.  I think the spec is actually missing this at the level of `countBy`. It specifies using the exception for keywords used in the remained of Query by Method Name that aren't supported but doesn't currently say that method name prefixes like `count` can be unsupported.

The `select` clause is optional in `select` statements. When it is missing, the query returns the queried entity.

==== Set clause

The `set` clause, with syntax given by `set_clause`, specifies a list of updates to fields of the queried entity. For each record which satisfies the restriction imposed by the `where` clause, and for each element of the list, the scalar expression is evaluated and assigned to the entity field identified by the path expression.

==== Order clause

The `order` clause (or `order by` clause), with syntax given by `orderby_clause`, specifies a lexicographic order for the query results, that is, a list of entity fields used to sort the records which satisfy the restriction imposed by the `where` clause. The keywords `asc` and `desc` specify that a given field should be sorted in ascending or descending order respectively; when neither is specified, ascending order is the default.

> The `order` clause is always optional. When it is missing, and when no sort criteria are given as arguments to a parameter of the repository method, the order of the query results is undefined, and may not be deterministic.

Reviewer 2: what happens when the db doesn't support `order`?

Author: Jakarta Data already has lots of ways to request sorting, so whatever happens in those other cases. Not sure where that's specified.

Reviewer: We don't have a guarantee for all fields in NoSQL; for example, Cassandra, when it works, will be for the primary key.
HBase does not support sorting at all; this also happens with key-value databases in the majority of scenarios.

At NoSQL, when the NoSQL database does not support I would expect the unsupportedoperationexception on those scenarios.

I mean, not only for Sorting, but even to operator.

For example, Cassandra has not support to **Or** operator.

Author: @graemerocher @otaviojava please review d127e7731db03a209101278ace53c2cf13933d41

Author: Another wrinkle here is that the apparently providers must raise `MappingException` when they can't do key-based pagination. That's not intuitive to me at all. It doesn't seem like a `MappingException` from what I can see.

Reviewer 3: > Another wrinkle here is that the apparently providers must raise `MappingException` when they can't do key-based pagination. That's not intuitive to me at all. It doesn't seem like a `MappingException` from what I can see.

We can change that to `UnsupportedOperationException`, too.

Author: Please see a4157bf2c0b9b0add9705792da069be07c0154e3.

This should actually be `DataException`, as per our recent discussion.

==== Logical operators

The logical operators are `and`, `or`, and `not`.

- An `and` operator expression is satisfied if and only if both its operands are satisfied.
- An `or` operator expression is satisfied if and only if at least one of its operands are satisfied.
- A `not` operator expression is never satisfied if its operand _is_ satisfied.

This specification leaves undefined the interpretation of the `not` operator when its operand _is not_ satisfied.

CAUTION: A compliant implementation of JDQL might feature SQL/JPQL-style ternary logic, where `not n > 0` is an unsatisfied logical expression when `n` evaluates to null, or it might feature binary logic where the same expression is considered satisfied. Application programmers should take great care when using the `not` operator with scalar expressions involving `null` values.

=== Clauses

Each JDQL statement is built from a sequence of _clauses_. The beginning of a clause is identified by a keyword: `from`, `where`, `select`, `set`, or `order`.

There is a logical ordering of clauses, reflecting the order in which their effect must be computed by the datastore:

1. `from`
2. `where`,
3. `select` or `set`,

> 4. `order`.

Reviewer 3: Are `select` and `order` reversed here?  In the case where `select` is reducing to an entity attribute, it seems like `order` would need to be done first to consider the various other entity attributes to order by.  For example,

`SELECT productId FROM Product ORDER BY price DESC, name ASC, productId ASC`

Author: In the full grammar for JPQL we have:

```
select_item ::= select_expression [[AS] result_variable]
select_expression ::=
    single_valued_path_expression |
    scalar_expression |
    aggregate_expression |
    identification_variable |
    OBJECT(identification_variable) |
    constructor_expression

...

orderby_item ::=
    state_field_path_expression |
    general_identification_variable |
    result_variable
    [ASC | DESC]
    [NULLS {FIRST | LAST}]
```

So you may order by any item in the `select` list, including complicated scalar expressions.

So yeah, it's correct.

Reviewer 3: Good point, we need to remain consistent with JPQL even though we don't provide the additional capabilities here

NOTE: This specification should not be interpreted to mandate an inefficient implementation of query language constructs in cases where the native behavior of the database varies from Java in such minor ways. That said, portability between Jakarta Data providers is maximized when their behavior is closest to the Java language.

Since a field of an entity may be null, a JDQL expression may evaluate to a null value.

=== Lexical structure

Lexical analysis requires recognition of the following token types:

- keywords (reserved identifiers),
- regular identifiers,
- named and ordinal parameters,
- operators and punctuation characters,
- literal strings, and
- integer and decimal number literals.

==== Identifiers and keywords

An _identifier_ is any legal Java identifier which is not a keyword. Identifiers are case-sensitive: `hello`, `Hello`, and `HELLO` are distinct identifiers.

In the JDQL grammar, identifiers are labelled with the `IDENTIFIER` token type.

> The following identifiers are _keywords_: `select`, `update`, `set`, `delete`, `from`, `where`, `order`, `by`, `asc`, `desc`, `not`, `and`, `or`, `between`, `like`, `in`, `null`, `local`, `true`, `false`. In addition, every reserved identifier listed in section 4.4.1 of the Jakarta Persistence specification version 3.2 is also considered a reserved identifier. Keywords and other reserved identifiers are case-insensitive: `null`, `Null`, and `NULL` are three ways to write the same keyword.

Author: Mmmmm. Hrrrrm. The assumption here is that section numbers are much more likely to change than section titles, which is _probably_ true ... but I dunno, a single-word doesn't seem like a wonderful sort of reference. Not sure if this is a good idea or not.

- The `right()` function is assigned the type `java.lang.String`, and evaluates to a suffix of the string to which its first argument evaluates. The length of the suffix is given by the integer value to which its second argument evaluates. The first argument must be of type `java.lang.String`, and the second argument must be of integral numeric type.

When any argument expression of any function call evaluates to a null value, the whole function call evaluates to null.

NOTE: These functions cannot be emulated on every datastore. When a function cannot be reasonably emulated via the native query capabilities of the database, a JDQL implementation is not required to provide the function.

If the JDQL implementation does not support a one of the standard functions explicitly listed above, it must throw `UnsupportedOperationException` when the function name occurs in a query. Alternatively, the Jakarta Data provider is permitted to reject a repository method declaration at compilation time if its `@Query` annotation uses an unsupported function.

> NOTE: On the other hand, an implementation of JDQL might provide additional built-in functions, and might even allow invocation of user-defined functions. Section 4.7 of the Jakarta Persistence specification defines a set of functions that all JPQL implementations are required to provide, including `concat`, `substring`, `trim`, `locate`, `ceiling`, `floor`, `exp`, `ln`, `mod`, `power`, `round`, `sign`, `sqrt`, `cast`, `extract`, `coalesce`, and `nullif`. JDQL implementations are encouraged to support any of these functions which are reasonably implementable.

Reviewer 3: The following could be done to remain correct if the sections are renumbered in Jakarta Persistence.
```suggestion
NOTE: On the other hand, an implementation of JDQL might provide additional built-in functions, and might even allow invocation of user-defined functions. The section of the Jakarta Persistence specification titled "Scalar Expressions" defines a set of functions that all JPQL implementations are required to provide, including `concat`, `substring`, `trim`, `locate`, `ceiling`, `floor`, `exp`, `ln`, `mod`, `power`, `round`, `sign`, `sqrt`, `cast`, `extract`, `coalesce`, and `nullif`. JDQL implementations are encouraged to support any of these functions which are reasonably implementable.
```

The four numeric operators may also be applied to an operand of wrapper type, for example, to `java.lang.Integer` or `java.lang.Double`. In this case, the operator expression is assigned a wrapper type, and evaluates to a null value when either of its operands evaluates to a null value. When both operands are non-null, the semantics are identical to the semantics of an operator expression involving the corresponding primitive types.

The four numeric operators may also be applied to operands of type `java.math.BigInteger` or `java.math.BigDecimal`.

A numeric operator expression is evaluated according to the native semantics of the database. In translating an operator expression to the native query language of the database, a Jakarta Data provider is encouraged, but not required, to apply reasonable transformations so that evaluation of the expression more closely mimics the semantics of the Java language.

==== Numeric types and numeric type promotion

> The type assigned to an operator expression depends on the types of its operand expression, which need not be identical. The rules for numeric promotion are given in section 4.7 of the Jakarta Persistence specification version 3.2:

Reviewer 3: ```suggestion
The type assigned to an operator expression depends on the types of its operand expression, which need not be identical. The rules for numeric promotion are given in the "Scalar Expressions" section of the Jakarta Persistence specification:
```

- equality and inequality operator expressions, and
- logical operator expressions.

The syntax for conditional expressions is given by the `conditional_expression` rule. Within a conditional expression, parentheses indicate grouping.

==== Null comparisons

A `null` comparison, with syntax given by `null_comparison_expression` is satisfied when:

- the `not` keyword is missing, and its operand evaluates to a null value, or
- the `not` keyword occurs, and its operand evaluates to any non-null value.

==== In expressions

An `in` expression, with syntax given by `in_expression` is satisfied when its leftmost operand evaluates to a non-null value, and:

- the `not` keyword is missing, and any one of its parenthesized operands evaluates to the same value as its leftmost operand, or
- the `not` keyword occurs, and none of its parenthesized operands evaluate to the same value as its leftmost operand.

All operands must have the same type.

==== Between expressions

A `between` expression, with syntax given by `between_expression` is satisfied when its operands all evaluate to non-null values, and, if the `not` keyword is missing, its left operand evaluates to a value which is:

> - larger than or equal to the value take by its middle operand, and

Reviewer 3: ```suggestion
- larger than or equal to the value taken by its middle operand, and
```

- smaller than or equal to the value taken by its right operand.

Or, if the `not` keyword occurs, the left operand must evaluate to a value which is:

> - strictly smaller than to the value take by its middle operand, or

Reviewer 3: ```suggestion
- strictly smaller than to the value taken by its middle operand, or
```

- Binary data is represented in JDQL as the type `byte[]`. Binary data is in general ordered lexicographically with respect to the constituent bytes. However, since this ordering is rarely meaningful, this specification does not require implementations of JDQL to respect it.

- This specification does not define an order for the sorting of Java `enum` values, which is provider-dependent. A programming model for entity classes might allow control over the order of `enum` values. For example, Jakarta Persistence allows this via the `@Enumerated` annotation.

- This specification does not define an order for UUID values, which is provider-dependent.

==== Logical operators

The logical operators are `and`, `or`, and `not`.

- An `and` operator expression is satisfied if and only if both its operands are satisfied.

> - An `or` operator expression is satisfied if and only if at least one of its operands are satisfied.

Reviewer 3: ```suggestion
- An `or` operator expression is satisfied if at least one of its operands is satisfied.
```

NOTE: The syntax of the `update` statement is irregular, with the `from` keyword implied. That is, the syntax _should_ be `update from Entity`, but for historical reasons it is simply `update Entity`.

The `from` clause is optional in `select` statements. When it is missing, the queried entity is determined by the return type of the repository method, or, if the return type is not an entity type, by the primary entity type of the repository.

For example, this repository method declaration:

[source,java]
----
@Query("where title like :title")
List<Book> booksByType(String title);
----

is equivalent to:

[source,java]
----
@Query("from Book where title like :title")
List<Book> booksByType(String title);
----

==== Where clause

The `where` clause, with syntax given by `where_clause`, specifies a conditional expression used to restrict the records returned, deleted, or updated by the query. Only records for which the conditional expression is satisfied are returned, deleted, or updated.

> The `where` clause is always optional. When it is missing, there is no restriction, and, and all records are of the queried entity type are returned, deleted, or updated.

Reviewer 3: ```suggestion
The `where` clause is always optional. When it is missing, there is no restriction, and all records of the queried entity type are returned, deleted, or updated.
```

The `select` clause is optional in `select` statements. When it is missing, the query returns the queried entity.

==== Set clause

The `set` clause, with syntax given by `set_clause`, specifies a list of updates to fields of the queried entity. For each record which satisfies the restriction imposed by the `where` clause, and for each element of the list, the scalar expression is evaluated and assigned to the entity field identified by the path expression.

==== Order clause

The `order` clause (or `order by` clause), with syntax given by `orderby_clause`, specifies a lexicographic order for the query results, that is, a list of entity fields used to sort the records which satisfy the restriction imposed by the `where` clause. The keywords `asc` and `desc` specify that a given field should be sorted in ascending or descending order respectively; when neither is specified, ascending order is the default.

> The `order` clause is always optional. When it is missing, and when no sort criteria are given as arguments to a parameter of the repository method, the order of the query results is undefined, and may not be deterministic.

Reviewer 3: ```suggestion
The `order` clause is always optional. When it is missing, and when no sort criteria are given as arguments to a parameter of the repository method, the order of the query results is undefined, and might not be deterministic.
```

The `order` clause (or `order by` clause), with syntax given by `orderby_clause`, specifies a lexicographic order for the query results, that is, a list of entity fields used to sort the records which satisfy the restriction imposed by the `where` clause. The keywords `asc` and `desc` specify that a given field should be sorted in ascending or descending order respectively; when neither is specified, ascending order is the default.

> By default, the order of values of a Java enumerated type is determined by the integer returned by `ordinal()`. A programming model for entity classes might allow control over the order of `enum` values. For example, Jakarta Persistence allows this via the `@Enumerated` annotation.

Reviewer: Are you sure that the `ordinal` should be the default?

If you include a new element or swap it, it will generate issues with data integrity.

When comparing String and `ordinal`, the unique benefit that came to my mind about ordinal is performance once it is smaller.

But, readability, data integrity, and flexibility I would prefer as String.

At the NoSQL side, we mostly use the `name`method, some samples:

- Couchbase
- Amazon DynamoDB
- RethinkDB
- Elasticsearch
- Cassandra
- Redis

Reviewer 3: This is an unfortunate situation where neither option is ideal. Ordering on `name` allows inserting new enum values to be a compatible change, but the user has no control over the desired ordering if they don't want alphabetical. Ordering on `ordinal` gives the user that control, but doesn't allow for inserting additional enum values at a later point without migrating existing data.  Given what Otavio brought up about some of the prominent NoSQL databases having the opposite default as Jakarta Persistence, we will need to define this as being provider-specific. I think it would be fine to state what the default is for Jakarta Persistence or mention the jakarta.persistence.Enumerated annotation that documents what the default is for Jakarta Persistence.  Maybe this would be better done in the JavaDoc.  I can look into that separately.

Author: OK, that's fine by me.
